components/p3a/brave_p3a_service.cc

```diff
@@ -0,0 +1,362 @@
+/* Copyright 2019 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/p3a/brave_p3a_service.h"
+
+#include <memory>
+#include <string>
+
+#include "base/command_line.h"
+#include "base/i18n/timezone.h"
+#include "base/metrics/histogram_samples.h"
+#include "base/metrics/metrics_hashes.h"
+#include "base/metrics/statistics_recorder.h"
+#include "base/rand_util.h"
+#include "base/strings/string_number_conversions.h"
+#include "base/task/post_task.h"
+#include "base/trace_event/trace_event.h"
+#include "brave/browser/brave_stats_updater_util.h"
+#include "brave/browser/version_info.h"
+#include "brave/common/pref_names.h"
+#include "brave/components/brave_prochlo/prochlo_message.pb.h"
+#include "brave/components/p3a/brave_p3a_log_store.h"
+#include "brave/components/p3a/brave_p3a_scheduler.h"
+#include "brave/components/p3a/brave_p3a_switches.h"
+#include "brave/components/p3a/brave_p3a_uploader.h"
+#include "brave/components/p3a/pref_names.h"
+#include "brave/vendor/brave_base/random.h"
+#include "chrome/browser/browser_process.h"
+#include "chrome/browser/first_run/first_run.h"
+#include "components/prefs/pref_registry_simple.h"
+#include "components/prefs/pref_service.h"
+#include "content/public/browser/browser_task_traits.h"
+#include "services/network/public/cpp/shared_url_loader_factory.h"
+#include "third_party/metrics_proto/reporting_info.pb.h"
+
+namespace brave {
+
+namespace {
+
+constexpr char kLastRotationTimeStampPref[] = "p3a.last_rotation_timestamp";
+
+constexpr char kDefaultUploadServerUrl[] = "https://p3a.brave.com/";
+
+constexpr uint64_t kDefaultUploadIntervalSeconds = 60 * 60;  // 1 hour.
+
+// TODO(iefremov): Provide moar histograms!
+// Whitelist for histograms that we collect. Will be replaced with something
+// updating on the fly.
+constexpr const char* kCollectedHistograms[] = {
+    "Brave.P3A.SentAnswersCount",
+    "Brave.Sync.Status",
+    "DefaultBrowser.State",
+    "Brave.Importer.ImporterSource",
+    "Brave.Shields.UsageStatus",
+    // Do not gather detailed info regarding TOR usage for now.
+    // "Brave.Core.LastTimeTorUsed",
+    "Brave.Core.TorEverUsed",
+    "Brave.Core.LastTimeIncognitoUsed",
+    "Brave.Core.NumberOfExtensions",
+    "Brave.Core.BookmarksCountOnProfileLoad",
+    "Brave.Core.TabCount",
+    "Brave.Core.WindowCount",
+    "Brave.Rewards.WalletBalance",
+    "Brave.Rewards.AutoContributionsState",
+    "Brave.Rewards.TipsState",
+    "Brave.Rewards.AdsState",
+    "Brave.Uptime.BrowserOpenMinutes",
+    "Brave.Welcome.InteractionStatus",
+};
+
+base::TimeDelta GetRandomizedUploadInterval(
+    base::TimeDelta max_upload_interval) {
+  return base::TimeDelta::FromSecondsD(
+      brave_base::random::Geometric(max_upload_interval.InSecondsF()));
```

Comment:
The parameter to random::Geometric is the _average_ delay between events, not the _maximum_.  (I am going to add a truncated geometric if there is a hard deadline for external reasons but I haven't yet; ideally, the hard deadline will be more than 4x the average interval.)  Is this supposed to be a hard deadline, or is this supposed to be a target for the average interval?

Reply:
@riastradh-brave Good catch, the variable name is misleading - should be `average_upload_interval_`.

Reply:
renamed